### PR TITLE
Fix module type functor

### LINKED ISSUE #671: No break when functor body is too long
As reported in #663 

```ocaml
  module type KV_MAKER = functor
    (G: Irmin_git.G)
    (C: Irmin.Contents.S) ->
    S with type key = string list
       and type step = string
       and type contents = C.t
       and type branch = string
       and module Git = G
```
is formatted into:

```ocaml
  module type KV_MAKER = functor (G : Irmin_git.G) (C : Irmin.Contents.S) -> S
                                                                             with type
                                                                             key =
                                                                               string
                                                                               list
                                                                              and type
                                                                             step =
                                                                               string
                                                                              and type
                                                                             contents =
                                                                               C
                                                                               .t
                                                                              and type
                                                                             branch =
                                                                               string
                                                                              and module Git = G
```

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #671 
Changing the formatting of modules is really tricky, we definitely need to redesign the whole function.

### CODE REVIEW

File: test/passing/shortcut_ext_attr.ml [hunk 1]
```diff
@@ -79,7 +79,8 @@ module type S = functor [@foo1]
   (M : S)
   -> functor
   (_ : (module type of M) [@foo2])
-  -> sig end [@foo3]
+  -> sig 
+  end [@foo3]
```
Comment (samoht) [inline test/passing/shortcut_ext_attr.ml]: this break looks a bit suspicious

Comment (gpetiot) [reply test/passing/shortcut_ext_attr.ml]: fixed

**Comment by gpetiot** [comment]
The regressions are all fixed, here is the diff with test_branch:

```ocaml
diff --git a/infer/src/absint/AbstractInterpreter.ml b/infer/src/absint/AbstractInterpreter.ml
index 046316415..3764f2157 100644
--- a/infer/src/absint/AbstractInterpreter.ml
+++ b/infer/src/absint/AbstractInterpreter.ml
@@ -418,8 +418,8 @@ module MakeUsingWTO (TransferFunctions : TransferFunctions.SIL) = struct
   let compute_post ?(do_narrowing = false) = make_compute_post ~exec_cfg_internal ~do_narrowing
 end
 
-module type Make = functor (TransferFunctions : TransferFunctions.SIL) -> S
-                                                                          with module TransferFunctions = TransferFunctions
+module type Make = functor (TransferFunctions : TransferFunctions.SIL) ->
+  S with module TransferFunctions = TransferFunctions
 
 module MakeRPO (T : TransferFunctions.SIL) =
   MakeWithScheduler (Scheduler.ReversePostorder (T.CFG)) (T)
diff --git a/infer/src/absint/AbstractInterpreter.mli b/infer/src/absint/AbstractInterpreter.mli
index 42dc68364..b20428cb3 100644
--- a/infer/src/absint/AbstractInterpreter.mli
+++ b/infer/src/absint/AbstractInterpreter.mli
@@ -58,8 +58,8 @@ module type S = sig
   (** extract the state for a node id from the given invariant map *)
 end
 
-module type Make = functor (TransferFunctions : TransferFunctions.SIL) -> S
-                                                                          with module TransferFunctions = TransferFunctions
+module type Make = functor (TransferFunctions : TransferFunctions.SIL) ->
+  S with module TransferFunctions = TransferFunctions
 
 (** create an intraprocedural abstract interpreter from transfer functions using the reverse post-order scheduler *)
 module MakeRPO : Make
diff --git a/ocamldoc/odoc_gen.ml b/ocamldoc/odoc_gen.ml
index 29f22356e..6afef55bd 100644
--- a/ocamldoc/odoc_gen.ml
+++ b/ocamldoc/odoc_gen.ml
@@ -33,20 +33,20 @@ end
 
 module type Base_functor = functor (G : Base) -> Base
 
-module type Html_functor = functor (G : Odoc_html.Html_generator) -> Odoc_html
-                                                                     .Html_generator
+module type Html_functor = functor (G : Odoc_html.Html_generator) ->
+  Odoc_html.Html_generator
 
-module type Latex_functor = functor (G : Odoc_latex.Latex_generator) -> Odoc_latex
-                                                                        .Latex_generator
+module type Latex_functor = functor (G : Odoc_latex.Latex_generator) ->
+  Odoc_latex.Latex_generator
 
-module type Texi_functor = functor (G : Odoc_texi.Texi_generator) -> Odoc_texi
-                                                                     .Texi_generator
+module type Texi_functor = functor (G : Odoc_texi.Texi_generator) ->
+  Odoc_texi.Texi_generator
 
-module type Man_functor = functor (G : Odoc_man.Man_generator) -> Odoc_man
-                                                                  .Man_generator
+module type Man_functor = functor (G : Odoc_man.Man_generator) ->
+  Odoc_man.Man_generator
 
-module type Dot_functor = functor (G : Odoc_dot.Dot_generator) -> Odoc_dot
-                                                                  .Dot_generator
+module type Dot_functor = functor (G : Odoc_dot.Dot_generator) ->
+  Odoc_dot.Dot_generator
 
 type generator =
   | Html of (module Odoc_html.Html_generator)
diff --git a/ocamldoc/odoc_gen.mli b/ocamldoc/odoc_gen.mli
index a42594a2b..768cb0d76 100644
--- a/ocamldoc/odoc_gen.mli
+++ b/ocamldoc/odoc_gen.mli
@@ -30,20 +30,20 @@ module Base_generator : Base
 
 module type Base_functor = functor (P : Base) -> Base
 
-module type Html_functor = functor (G : Odoc_html.Html_generator) -> Odoc_html
-                                                                     .Html_generator
+module type Html_functor = functor (G : Odoc_html.Html_generator) ->
+  Odoc_html.Html_generator
 
-module type Latex_functor = functor (G : Odoc_latex.Latex_generator) -> Odoc_latex
-                                                                        .Latex_generator
+module type Latex_functor = functor (G : Odoc_latex.Latex_generator) ->
+  Odoc_latex.Latex_generator
 
-module type Texi_functor = functor (G : Odoc_texi.Texi_generator) -> Odoc_texi
-                                                                     .Texi_generator
+module type Texi_functor = functor (G : Odoc_texi.Texi_generator) ->
+  Odoc_texi.Texi_generator
 
-module type Man_functor = functor (G : Odoc_man.Man_generator) -> Odoc_man
-                                                                  .Man_generator
+module type Man_functor = functor (G : Odoc_man.Man_generator) ->
+  Odoc_man.Man_generator
 
-module type Dot_functor = functor (G : Odoc_dot.Dot_generator) -> Odoc_dot
-                                                                  .Dot_generator
+module type Dot_functor = functor (G : Odoc_dot.Dot_generator) ->
+  Odoc_dot.Dot_generator
+module type Man_functor = functor (G : Odoc_man.Man_generator) ->
+  Odoc_man.Man_generator
 
-module type Dot_functor = functor (G : Odoc_dot.Dot_generator) -> Odoc_dot
-                                                                  .Dot_generato
r
+module type Dot_functor = functor (G : Odoc_dot.Dot_generator) ->
+  Odoc_dot.Dot_generator
 
 (** Various ways to create a generator. *)
 type generator =
diff --git a/testsuite/tests/basic/sets.ml b/testsuite/tests/basic/sets.ml
index 2d8d7d587..b2ecb83dc 100644
--- a/testsuite/tests/basic/sets.ml
+++ b/testsuite/tests/basic/sets.ml
@@ -16,7 +16,8 @@ let _ =
   done
 
 module PowerSet
-    (BaseSet : Set.S) (SetOrd : functor (S : Set.S) -> Set.OrderedType) =
+    (BaseSet : Set.S) (SetOrd : functor (S : Set.S) ->
+      Set.OrderedType) =
   Set.Make (SetOrd (BaseSet))
 module IntSetSet = PowerSet (IntSet) ((functor (S : Set.S) -> S))
 
diff --git a/testsuite/tests/typing-recmod/t17ok.ml b/testsuite/tests/typing-recmod/t17ok.ml
index aa2afa02c..a088b05fd 100644
--- a/testsuite/tests/typing-recmod/t17ok.ml
+++ b/testsuite/tests/typing-recmod/t17ok.ml
@@ -24,10 +24,9 @@ end
 
 type 'a tree = E | N of 'a tree * 'a * 'a tree
 
-module Bootstrap2 (MakeDiet : functor (X : ORD) -> SET
-                                                   with type t = X.t tree
-                                                    and type elt = X.t) :
-  SET with type elt = int = struct
+module Bootstrap2 (MakeDiet : functor (X : ORD) ->
+  SET with type t = X.t tree and type elt = X.t) : SET with type elt = int =
+struct
   type elt = int
 
   module rec Elt : sig
diff --git a/testsuite/tests/typing-recmod/t22ok.ml b/testsuite/tests/typing-recmod/t22ok.ml
index aaa3b2094..6b0b2029e 100644
--- a/testsuite/tests/typing-recmod/t22ok.ml
+++ b/testsuite/tests/typing-recmod/t22ok.ml
@@ -292,8 +292,8 @@ module type HEAP = sig
   val deleteMin : heap -> heap
 end
 
-module Bootstrap (MakeH : functor (Element : ORDERED) -> HEAP
-                                                         with module Elem = Element)
+module Bootstrap (MakeH : functor (Element : ORDERED) ->
+  HEAP with module Elem = Element)
 (Element : ORDERED) : HEAP with module Elem = Element = struct
   module Elem = Element
 
diff --git a/testsuite/tests/typing-signatures/els.ml b/testsuite/tests/typing-signatures/els.ml
index 865bb0210..d34faa3ff 100644
--- a/testsuite/tests/typing-signatures/els.ml
+++ b/testsuite/tests/typing-signatures/els.ml
@@ -110,11 +110,8 @@ module type BARECODE = sig
 end
 
 module USERCODE (X : TYPEVIEW) = struct
-  module type F = functor (C : CORE with type V.usert = X.combined) -> BARECODE
-                                                                       with type 
-                                                                       state :=
-                                                                         C.V
-                                                                         .state
+  module type F = functor (C : CORE with type V.usert = X.combined) ->
+    BARECODE with type state := C.V.state
 end
```

**Comment by samoht** [comment]
This one is a bit strange:
```
-    (BaseSet : Set.S) (SetOrd : functor (S : Set.S) -> Set.OrderedType) =
+    (BaseSet : Set.S) (SetOrd : functor (S : Set.S) ->
+      Set.OrderedType) =
```

File: src/Fmt_ast.ml [hunk 2]
```diff
@@ -3127,8 +3127,14 @@ and fmt_module_type c ({ast= mty} as xmty) =
             $ fmt_attributes c ~pre:(fmt " ") ~key:"@" pmty_attributes
             $ fmt "@;<1 2>"
             $ list xargs "@;<1 2>" fmt_arg
-            $ fmt "@;<1 2>-> " $ Option.call ~f:blk.pro )
-      ; epi= Some (Option.call ~f:blk.epi $ Cmts.fmt_after c pmty_loc) }
+            $ fmt "@;<1 2>->"
+            $ fmt_if (Option.is_some blk.pro) " "
+            $ Option.call ~f:blk.pro )
```
[the remark this answers is not in the document]
Comment (jberdine) [reply src/Fmt_ast.ml]: or add a `?pro` arg to `opt` and write `opt ~pro:(fmt " ") blk.pro`

[the remark this answers is not in the document]
Comment (Julow) [reply src/Fmt_ast.ml]: `opt` takes a function and is in the "primitive containers" section. What about a `fmt_opt` in the "conditional formatting" section ?

**Comment by gpetiot** [comment]
> This one is a bit strange:
> 
> ```ocaml
> -    (BaseSet : Set.S) (SetOrd : functor (S : Set.S) -> Set.OrderedType) =
> +    (BaseSet : Set.S) (SetOrd : functor (S : Set.S) ->
> +      Set.OrderedType) =
> ```
> 
> Is it because the line becomes longer than 80 chars?

It was a regression, it is fixed with my last commit.
The rest of the diff of test_branch is unchanged.

File: src/Fmt_ast.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (gpetiot) [reply src/Fmt_ast.ml]: I chose Jules' suggestion, maybe we can change Fmt functions' signatures in another PR (and `opt` still needs a function as a second parameter (i.e. Fn.id))

**Pull request merged**